src/Interpreters/TreeRewriter.cpp

```diff
@@ -509,6 +509,186 @@ void setJoinStrictness(ASTSelectQuery & select_query, JoinStrictness join_defaul
     out_table_join = table_join;
 }
 
+class DNF
+{
+    bool node_added = false;
+
+    void normTree(ASTPtr node)
+    {
+        auto *function = node->as<ASTFunction>();
+
+        // LOG_TRACE(&Poco::Logger::get("TreeRewrite"), "top of normTree:  {}", node->dumpTree());
+        if (function && function->children.size() == 1)
```

Comment:
I think that the assumption "we need to throw error if function with more that one child met" is correct. Well spotted!
I planned to do it but did not dare.
We need this check to survive in case of garbage requests (e.g. generated by fuzzer), for example OR without children at all.
I did not add throwing error because the problem is not new, it is reproducible with ANDs, and ClickHouse somehow processes such queries without a complain.
So, yes, I confirm that we have a problem here.